### PR TITLE
protoc-gen-flowtypes: add optional field support

### PR DESCRIPTION
**Pull request by tmc**
this refactors and introduce the option to mark message fields as optional via https://github.com/mwitkow/go-proto-validators

### CODE REVIEW

File: protoc-gen-flowtypes/README.md [hunk 1]
```diff
@@ -36,11 +36,11 @@ message SearchResponse {
 
 # example use:
 ```sh
-$ protoc -I. --flowtypes_out=. simple.proto
-$ cat simpleTypes.js
+$ protoc -I. -I${GOPATH}/src --flowtypes_out=. simple.proto
+$ cat simple_types.js
```
Comment (alextoombs) [inline protoc-gen-flowtypes/README.md:40]: isn't `SimpleTypes.js` more ES6-canonical?

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go [hunk 2]
```diff
@@ -6,73 +6,97 @@ import (
...
 
-func (t *namedFlowType) FlowType() string {
-	return t.Type.FlowType()
-	return fmt.Sprintf("%s = %s", t.Name, t.Type.FlowType())
+func (r repeatedType) FlowType() string { return fmt.Sprintf("Array<%s>", r.Flower.FlowType()) }
+
+type namedType struct {
+	Flower
+	name     string
+	required bool // TODO(tmc): move to interface
```
Comment (alextoombs) [inline protoc-gen-flowtypes/genflowtypes/genflowtypes.go]: This is already on the interface definition, right?

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go [hunk 3]
```diff
@@ -6,73 +6,97 @@ import (
...
-	return t.Name
+
+func (t *namedType) Name() string {
+	return t.name
 }
 
 type objectFlowType struct {
-	Fields  []NamedFlowType
-	Options Options
+	Fields   []NamedFlower
```
Comment (alextoombs) [inline protoc-gen-flowtypes/genflowtypes/genflowtypes.go:70]: Fields are always named in an object? Is that the most prevalent used of name flow types? (for field names)

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go [hunk 4]
```diff
@@ -6,73 +6,97 @@ import (
...
-		fields = append(fields, fmt.Sprintf("  %s%s: %s", f.FlowTypeName(), optionalIndicator, f.FlowType()))
+		fields = append(fields, fmt.Sprintf("  %s%s: %s", f.Name(), optionalIndicator, f.FlowType()))
 	}
 	return fmt.Sprintf("{\n%s\n}", strings.Join(fields, ",\n"))
 }
 
-func (cfg Options) fqmnToType(fqmn string, registry *descriptor.Registry) (FlowType, error) {
+func (t *objectFlowType) IsRequired() bool { return t.required }
+
+func (cfg Options) fqmnToType(fqmn string, registry *descriptor.Registry) (Flower, error) {
```
Comment (alextoombs) [inline protoc-gen-flowtypes/genflowtypes/genflowtypes.go:89]: fqmn?

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go [hunk 5]
```diff
@@ -6,73 +6,97 @@ import (
...
 	m, err := registry.LookupMsg("", fqmn)
 	if err != nil {
 		return nil, err
 	}
 	return cfg.messageToFlowType(m, registry)
 }
 
-func (cfg Options) fieldToType(f *descriptor.Field, reg *descriptor.Registry) (NamedFlowType, error) {
+func (cfg Options) fieldToType(f *descriptor.Field, reg *descriptor.Registry, required bool) (NamedFlower, error) {
```
Comment (alextoombs) [inline protoc-gen-flowtypes/genflowtypes/genflowtypes.go:97]: Isn't the required nature already carried on `f`?

**Review by alextoombs — APPROVED** [review]
Just some questions; overall seems good.

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go [hunk 6]
```diff
@@ -126,28 +150,57 @@ func (cfg Options) fieldToType(f *descriptor.Field, reg *descriptor.Registry) (N
...
 	if f.GetLabel() == pbdescriptor.FieldDescriptorProto_LABEL_REPEATED {
-		fieldType = repeatedFlowType{fieldType}
+		fieldType = newRepeatedFlowType(fieldType, required)
 	}
-	return &namedFlowType{Name: f.GetName(), Type: fieldType}, nil
+	return &namedType{Flower: fieldType, name: f.GetName(), required: required}, nil
 }
 
-func (cfg Options) messageToFlowType(m *descriptor.Message, reg *descriptor.Registry) (FlowType, error) {
+// this is a hack to use the pbdescptor from golang/proto instead of gogo/proto
```
Comment (johanbrandhorst) [inline protoc-gen-flowtypes/genflowtypes/genflowtypes.go]: pbdescriptor

File: protoc-gen-flowtypes/README.md (hunk 1 above)
Comment (tmc) [reply protoc-gen-flowtypes/README.md:40]: this is overridable by the user and unrelated to this change (just fixing docs)

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go (hunk 2 above)
Comment (tmc) [reply protoc-gen-flowtypes/genflowtypes/genflowtypes.go]: yep, stale.

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go (hunk 4 above)
Comment (tmc) [reply protoc-gen-flowtypes/genflowtypes/genflowtypes.go:89]: fully qualified message name

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go (hunk 3 above)
Comment (tmc) [reply protoc-gen-flowtypes/genflowtypes/genflowtypes.go:70]: yes I believe so, I'm not sure about the relative counts of those in the wild, why?

**Review by johanbrandhorst — APPROVED** [review]
LGTM

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go (hunk 5 above)
Comment (tmc) [reply protoc-gen-flowtypes/genflowtypes/genflowtypes.go:97]: we evaluate required vs not in messageToFlowType

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go (hunk 6 above)
Comment (tmc) [reply protoc-gen-flowtypes/genflowtypes/genflowtypes.go]: thanks

**Pull request merged**